text/0000-global-packages.md

```diff
@@ -0,0 +1,155 @@
+
+## Handling Global Libraries
+
+Since libraries without binaries are already supported by `npm i -g`, this approach also includes that support without needing any modifications.
+
+## `volta list`
+
+Since this change represents a significant update of the internal model for packages, it will require the output of `volta list` around packages to be redesigned. Specifically, packages will no longer be associated with a specific platform any more, so we should be able to simplify the output to only show the package version (and whether it is the default version or a project-local version).
+
+## Workflow Examples
+
+On a fresh install, a user selects the latest 12.* version of Node as their default:
+
+```bash
+> volta install node@12
+```
+
+Once this is complete, they are using `node@12` and don't have any global packages installed yet. Next, the user installs TypeScript 3.9 as a global:
+
+```bash
+> npm i -g typescript@3.9
+```
+
+Volta detects this installation and ensures that `npm` installs `typescript` inside of Volta's internal Node 12 bucket. It also notes the specific version (3.9.7 as of this writing) of `typescript` that is installed in the manifest of installed packages.
+
+Now, the user decides they actually want to work on the bleeding edge of Node, so they install Node 14:
+
+```bash
+> volta install node@14
+```
+
+After making Node available, Volta will check the manifest of global packages and see that we need to make sure `typescript@3.9.7` is available, so it will internally run `npm i -g typescript@3.9.7` under Node 14, making sure that it is installed into the Node 14 bucket. At the end of this process, the user can run `tsc` as a global just like they could under Node 12.
+
+While working in Node 14, the user installs `ember-cli@3.18.0` and also decides that they want to downgrade to `typescript@3.8.2`. Volta ensures that those changes are reflected in the Node 14 bucket, as well as keeping track of the active versions in the global manifest:
+
+```bash
+> npm i -g ember-cli@3.18.0
+> npm i -g typescript@3.8.2
+```
+
+Finally, the user decides that working on Node 14 isn't working out, and they switch back to Node 12:
+
+```bash
+> volta install node@12
+```
+
+After ensuring Node 12 is available, Volta will once again check the manifest of global packages against the packages that are currently set up in the Node 12 bucket. In this case, Volta detects that we need to install `ember-cli@3.18.0` and downgrade `typescript` to `3.8.2`, so it runs `npm` commands internally to ensure those changes are made. When the process is complete, the user has the exact same tools that they did on Node 14.
```

Comment:
I'm curious in this case:
After switching back to `node@12`, let's just assume `ember-cli@3.18.0` is not compatible with `node@12`, and its min requirement is `node>=14`:

1. What will happen to the global `ember-cli` after switching back to `node@12`? For me personally I would expect some message to tell me this issue at least, like:

```
success: installed and set node@12.16.2 (with npm@6.14.4) as default
Checking global installed packages...
- [success] typescript@3.8.2
- [fail] ember-cli@3.18.0
    ember-cli@3.18.0 is not compatible with node@12.16.2, it requires `node>=14`. Please try downgrade `ember-cli` or upgrade `node` in order to use it
```

2. Most of the "good" packages will define the node version requirement in `package.json` via `engine`, and I assume we will also use it to check the compatibility. However, what if a package `foo` requires `node>=14`, but  don't have the `engine` defined in `package.json`(or wrong version)? After switching back to `node@12`, do we have better ways to tell if `foo` would work with `node@12`? And if user fails to run `foo` with `node@12`, do we have a way to let user know what's going on?

Reply:
1. I was thinking something like that, yes, and I really like that output so we can easily show the status of the installed tools.

2. Outside of `engines`—or more generally, the `npm i -g` command failing—I'm not sure of a good way to determine if a given package is compatible with a specific Node version or not.